test/common/http/http2/http2_connection_fuzz_test.cc

```diff
@@ -0,0 +1,314 @@
+    data_[1] = (size >> 8) & 0xff;
+    data_[2] = size & 0xff;
+  }
+
+  void adjustPayloadSize() { setPayloadSize(data_.size() - HeaderSize); }
+
+  void setType(uint8_t type) { data_[3] = type; }
+  void setFlags(uint8_t flags) { data_[4] = flags; }
+
+  void makeClientStreamId() {
+    uint32_t id = (stream_id_ << 1) + 1;
+    if (data_.size() >= HeaderSize) {
+      std::memcpy(&data_[5], &id, sizeof(id));
+    }
+  }
+
+  void makeServerStreamId() {
+    uint32_t id = stream_id_ << 1;
+    if (data_.size() >= HeaderSize) {
+      std::memcpy(&data_[5], &id, sizeof(id));
+    }
+  }
+
+  void appendData(absl::string_view data) { data_.insert(data_.end(), data.begin(), data.end()); }
+
+  const absl::string_view payload() const {
+    const char* d = reinterpret_cast<const char*>(data_.data());
+    return absl::string_view(d, data_.size());
+  }
+
+private:
+  uint32_t stream_id_;
+  std::vector<uint8_t> data_;
+};
+
+void deflate_headers(nghttp2_hd_deflater* deflater, const test::fuzz::Headers& fragment,
+                     std::string& payload) {
+#define MAX_HDRS 32
+  nghttp2_nv h2hdrs[MAX_HEADERS];
+  uint8_t buf[MAX_HD_TABLE_BUF_SIZE];
+
+  unsigned n_headers = fragment.headers().size();
+  if (n_headers > MAX_HEADERS)
+    n_headers = MAX_HEADERS;
+  auto headers = fragment.headers();
+  for (unsigned i = 0; i < n_headers; i++) {
+    auto hv = headers.at(i);
+    size_t keylen = hv.key().size();
+    uint8_t* key = reinterpret_cast<uint8_t*>(alloca(keylen));
+    std::memcpy(key, hv.key().c_str(), keylen);
+
+    size_t valuelen = hv.value().size();
+    uint8_t* value = reinterpret_cast<uint8_t*>(alloca(valuelen));
+    std::memcpy(value, hv.value().c_str(), valuelen);
+
+    h2hdrs[i].name = key;
+    h2hdrs[i].namelen = keylen;
+    h2hdrs[i].value = value;
+    h2hdrs[i].valuelen = valuelen;
+    h2hdrs[i].flags = 0;
+  }
+  ssize_t len = nghttp2_hd_deflate_hd(deflater, buf, sizeof(buf), h2hdrs, n_headers);
+  if (len > 0)
+    payload.append(reinterpret_cast<char*>(buf), len);
+}
+
+Http2Frame pb_to_h2_frame(nghttp2_hd_deflater* deflater,
+                          const test::common::http::http2::Http2FrameOrJunk& pb_frame) {
+  if (pb_frame.has_h2frame()) {
+    // Need to encode headers
+    auto h2frame = pb_frame.h2frame();
+    uint32_t streamid = h2frame.streamid();
+    uint8_t flags = static_cast<uint8_t>(h2frame.flags() & 0xff);
+    uint8_t type;
+    std::string payload;
+    if (h2frame.has_data()) {
+      type = 0;
+      auto f = h2frame.data();
+      uint8_t padding_len = static_cast<uint8_t>(f.padding().size() & 0xff);
+      payload.push_back(static_cast<char>(padding_len));
+      payload.append(f.data());
+      payload.append(f.padding().c_str(), padding_len);
+    } else if (h2frame.has_headers()) {
+      type = 1;
+      auto f = h2frame.headers();
+      uint8_t padding_len = static_cast<uint8_t>(f.padding().size() & 0xff);
+      payload.push_back(static_cast<char>(padding_len));
```

Comment:
The padding length byte is supposed to only be added if the PADDED flag is set. Is it intentional that we always add the padding length byte, regardless of flag state?

Reply:
For the HEADERS and PUSH_PROMISE the code now adds padding depending on the flag. Unconstrained fuzzing is still happening due to the junk frames.